src/instrumentation/sdk.go

```diff
@@ -19,496 +19,98 @@ package instrumentation
-		for _, kv := range existingResArr {
-			keyValueArr := strings.Split(strings.TrimSpace(kv), "=")
-			if len(keyValueArr) != 2 {
+	found := false
+	for _, inst := range insts {
+		for _, injector := range i.injectorRegistry.GetInjectors() {
+			if injector.Language() != inst.Spec.Agent.Language {
 				continue
 			}
-			existingRes[keyValueArr[0]] = true
-		}
-	}
-
-	res := map[string]string{}
-	for k, v := range newrelic.Spec.Resource.Attributes {
-		if !existingRes[k] {
-			res[k] = v
-		}
-	}
-	k8sResources := map[attribute.Key]string{}
-	k8sResources[semconv.K8SNamespaceNameKey] = ns.Name
-	k8sResources[semconv.K8SContainerNameKey] = pod.Spec.Containers[index].Name
-	// Some fields might be empty - node name, pod name
-	// The pod name might be empty if the pod is created form deployment template
-	k8sResources[semconv.K8SPodNameKey] = pod.Name
-	k8sResources[semconv.K8SPodUIDKey] = string(pod.UID)
-	k8sResources[semconv.K8SNodeNameKey] = pod.Spec.NodeName
-	k8sResources[semconv.ServiceInstanceIDKey] = createServiceInstanceId(ns.Name, pod.Name, pod.Spec.Containers[index].Name)
-	i.addParentResourceLabels(ctx, newrelic.Spec.Resource.AddK8sUIDAttributes, ns, pod.ObjectMeta, k8sResources)
-	for k, v := range k8sResources {
-		if !existingRes[string(k)] && v != "" {
-			res[string(k)] = v
-		}
-	}
-	return res
-}
-
-func (i *sdkInjector) addParentResourceLabels(ctx context.Context, uid bool, ns corev1.Namespace, objectMeta metav1.ObjectMeta, resources map[attribute.Key]string) {
-	for _, owner := range objectMeta.OwnerReferences {
-		switch strings.ToLower(owner.Kind) {
-		case "replicaset":
-			resources[semconv.K8SReplicaSetNameKey] = owner.Name
-			if uid {
-				resources[semconv.K8SReplicaSetUIDKey] = string(owner.UID)
-			}
-			// parent of ReplicaSet is e.g. Deployment which we are interested to know
-			rs := appsv1.ReplicaSet{}
-			nsn := types.NamespacedName{Namespace: ns.Name, Name: owner.Name}
-			backOff := wait.Backoff{Duration: 10 * time.Millisecond, Factor: 1.5, Jitter: 0.1, Steps: 20, Cap: 2 * time.Second}
-
-			checkError := func(err error) bool {
-				return apierrors.IsNotFound(err)
-			}
-
-			getReplicaSet := func() error {
-				return i.client.Get(ctx, nsn, &rs)
-			}
-
-			// use a retry loop to get the Deployment. A single call to client.get fails occasionally
-			err := retry.OnError(backOff, checkError, getReplicaSet)
+			found = true
+			injector.ConfigureClient(i.client)
+			injector.ConfigureLogger(i.logger.WithValues("injector", injector.Language()))
+			i.logger.V(1).Info("injecting instrumentation into pod",
+				"agent_language", inst.Spec.Agent.Language,
+				"newrelic-namespace", inst.Namespace,
+				"newrelic-name", inst.Name,
+			)
+
+			var newPod corev1.Pod
+			var err error
+			func() {
+				defer func() {
+					if r := recover(); r != nil {
+						err = fmt.Errorf("panic: %v, stacktrace: %s", r, debug.Stack())
+					}
+				}()
+				newPod, err = injector.Inject(ctx, *inst, ns, pod)
+			}()
```

Comment:
I think there is a race condition here.  I don't think we should `Inject` inside a func() here since it does not wait for anything. So, the err below will mostly be `nil`. 

If you want to recover from panic in a loop, golang recommends to have the whole body of the loop inside a function so that defer can kick in properly e.g.

https://go.dev/play/p/NVOTA2z_XIs

Reply:
Same as https://go.dev/play/p/Ii77rO3VHlY but the above is recommended and we should not be using a common variable like `err` that is common to both funcs.

Reply:
I suggest not to worry about panic handling in this iteration of PR.

Reply:
Never mind, I think i saw a `go func()` somewhere. So, there is no race condition but I still think you should wrap the whole thing in func() inside the loop.

Reply:
The intent of handling panics is to handle each injectors possible panic separately, which is why I'm calling it that way in the loop.

You are correct that each injector is *not* called as a go routine, so there's no need to add synchronization primitives or panics in go routines. 

I'll break this up a bit.